explorer/interpreter/interpreter.cpp

```diff
@@ -1449,15 +1449,42 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
           CARBON_CHECK(args.size() == 1);
           CARBON_CHECK(act.pos() > 0);
           const auto* ptr = cast<PointerValue>(args[0]);
-          if (act.pos() == 1) {
-            CARBON_ASSIGN_OR_RETURN(
-                const auto* pointee,
-                this->heap_.Read(ptr->address(), exp.source_loc()));
-            return todo_.Spawn(std::make_unique<DestroyAction>(
-                arena_->New<LValue>(ptr->address()), pointee));
+          CARBON_ASSIGN_OR_RETURN(const auto* pointee,
+                                  heap_.Read(ptr->address(), exp.source_loc()));
+          if (const auto* class_value = dyn_cast<NominalClassValue>(pointee)) {
+            // Handle destruction from base class pointer.
+            const auto* child_class_value = *class_value->class_value_ptr();
+            bool is_subtyped = child_class_value != class_value;
+            const Address obj_addr = is_subtyped
+                                         ? ptr->address().DowncastedAddress()
+                                         : ptr->address();
+            if (is_subtyped) {
+              // Error if destructor is not virtual.
+              const auto& class_type =
+                  cast<NominalClassType>(class_value->type());
+              const auto& class_decl = class_type.declaration();
+              if ((*class_decl.destructor())->virt_override() ==
+                  VirtualOverride::None) {
+                return ProgramError(exp.source_loc())
+                       << "Deallocating a derived class from base class "
+                          "pointer requires a virtual destructor";
+              }
+            }
```

Comment:
```suggestion
            if (is_subtyped) {
              // Error if destructor is not virtual.
              const auto& class_type =
                  cast<NominalClassType>(class_value->type());
              const auto& class_decl = class_type.declaration();
              if ((*class_decl.destructor())->virt_override() ==
                  VirtualOverride::None) {
                return ProgramError(exp.source_loc())
                       << "Deallocating a derived class from base class "
                          "pointer requires a virtual destructor";
              }
            }
            const Address obj_addr = is_subtyped
                                         ? ptr->address().DowncastedAddress()
                                         : ptr->address();
```

obj_addr is mainly used below; I think it may be clearer if the virtual destructor check is done first.